apps/aecore/src/aec_tx_pool_sync.erl

```diff
@@ -0,0 +1,477 @@
+        {error, _} ->
+            epoch_sync:info("Peer connection died - TX-pool synchronization aborted!", []),
+            State#state{ local = not_synced }
+    end;
+handle_down(State = #state{ remote = Remotes }, Ref) ->
+    case lists:keyfind(Ref, #sync.peer_con_ref, Remotes) of
+        false ->
+            State;
+        Sync = #sync{ id = Id } ->
+            case repair_sync(Sync) of
+                {ok, NewSync} ->
+                    State#state{ remote = lists:keyreplace(Id, #sync.id, Remotes, NewSync) };
+                {error, _} ->
+                    epoch_sync:info("Peer connection died for remote - TX-pool sync aborted!", []),
+                    State#state{ remote = lists:keydelete(Id, #sync.id, Remotes) }
+            end
+    end.
+
+handle_local_action(State = #state{ local = {active, #sync{ id = Ref }} }, Action, Ref, Result) ->
+    handle_local_action(State, Action, Result);
+handle_local_action(State = #state{ remote = Remotes }, tree, Ref, Result) ->
+    case lists:keyfind(Ref, #sync.id, Remotes) of
+        false -> %% ignore stale result
+            State;
+        Sync = #sync{} ->
+            case Result of
+                {ok, Tree} ->
+                    Sync1 = Sync#sync{ tree = Tree },
+                    State#state{ remote = lists:keyreplace(Ref, #sync.id, Remotes, Sync1) };
+                {error, _} ->
+                    State#state{ remote = lists:keydelete(Ref, #sync.id, Remotes) }
+            end
+    end.
+
+handle_local_action(State = #state{ local = {active, Sync} }, Action, {error, _} = Err) ->
+    epoch_sync:debug("Local TX-pool sync action ~p failed (~p)", [Action, Err]),
+    do_local_action(Sync, {finish, false}),
+    State;
+handle_local_action(State = #state{ local = {active, Sync} }, Action, Res) ->
+    case {Action, Res} of
+        {tree, {ok, Tree}} ->
+            do_local_action(Sync, init),
+            State#state{ local = {active, Sync#sync{ tree = Tree }} };
+        {init, ok} ->
+            RootHash = aeu_mp_trees:root_hash(Sync#sync.tree),
+            do_local_action(Sync, {unfold, [{node, <<>>, RootHash}], 0}),
+            State#state{ local = {active, Sync#sync{ data = [] }} };
+        {unfold, {ok, tree_not_ready, OldUnfolds, Delay}} when Delay < 10000 ->
+            %% Remote tree is not computed, retry with delay
+            epoch_sync:debug("Remote tree temporay failure, retrying in ~p ms",
+                             [Delay + 1000]),
+            do_local_action(Sync, {unfold, OldUnfolds, Delay + 1000}),
+            State;
+        {unfold, {ok, tree_not_ready, _OldUnfolds, _Delay}} ->
+            epoch_sync:debug("Remote tree permanent failure, aborting", []),
+            do_local_action(Sync, {finish, false}),
+            State;
+        {unfold, {ok, NewUnfolds}} ->
+            #sync{ data = OldGets } = Sync,
+            {NewUnfolds1, NewGets} = analyze_unfolds(NewUnfolds, Sync#sync.tree),
```

Comment:
This may terminate abnormally because [deserialize_unfolds](https://github.com/aeternity/epoch/pull/1245/files#diff-0c07dd015037eb446bbd6e7298a61532R290) may return, among the items in the returned list, `{error, _}` (as input from network, and not deeply sanitized because only sanitized up to simple [deserialization](https://github.com/aeternity/epoch/pull/1245/files#diff-c1692ca9a93deb9e4ca76237ebbc01fdR217)). Expected?

Reply:
I actually never expect deserialize_unfold to return {error, _} since all data it is called on is produced by serialize_unfold which is (or at least should be) symmetric...

Reply:
Will make the error case explicit.